crypto/core_namemap.c

```diff
@@ -163,19 +163,31 @@ int ossl_namemap_name2num(const OSSL_NAMEMAP *namemap, const char *name)
     return number;
 }
 
-/* TODO: Optimize to avoid strndup() */
 int ossl_namemap_name2num_n(const OSSL_NAMEMAP *namemap,
                             const char *name, size_t name_len)
 {
-    char *tmp;
-    int ret;
+    int number = 0;
+    HT_VALUE *val;
+    NAMENUM_KEY key;
 
-    if (name == NULL || (tmp = OPENSSL_strndup(name, name_len)) == NULL)
+#ifndef FIPS_MODULE
+    if (namemap == NULL)
+        namemap = ossl_namemap_stored(NULL);
+#endif
+
+    if (namemap == NULL)
         return 0;
 
-    ret = ossl_namemap_name2num(namemap, tmp);
-    OPENSSL_free(tmp);
-    return ret;
+    HT_INIT_KEY(&key);
+    HT_SET_KEY_STRING_CASE_N(&key, name, name, name_len);
```

Comment:
without the strdup here, we risk converting the passed in name to uppercase, which may behave improperly, if the caller expects the string to be unmodified in its call.

That said, I like the idea.  What you could do instead is modify the HT_SET_KEY_STRING_CASE macro to first copy the string into the key buffer, and then run ossl_ht_strcase on the copy in the key buffer.  That would eliminate the need to duplicate the string without risking modifications to the caller

Reply:
I don't think this is a problem because it is exactly how HT_SET_KEY_STRING_CASE was written before. 
The declaration is `ossl_ht_strcase(char *tgt, const char *src, int len)`. The `const char *src` means that `ossl_ht_strcase` will not modify the passed in name, and so this isn't a problem.

And anyways, if we do that, then we might as well just leave the code as it is now and just remove the `TODO` comment

Reply:
yeah, you're right, ossl_ht_strcase copies the input string to the buffer, while doing the case conversion, we're ok here, sorry for the noise.  

Kind of begs the question though why we ever did the strdup in the first place.  Wierd.